include/swift/AST/Expr.h

```diff
@@ -5578,7 +5579,11 @@ class KeyPathExpr : public Expr {
   /// components from the argument array.
   void resolveComponents(ASTContext &C,
                          ArrayRef<Component> resolvedComponents);
-  
+
+  /// Indicates if the key path expression is composed by a single invalid
+  /// component. e.g. missing component `\Root`
+  bool hasSingleInvalidComponent() const { return HasSingleInvalidComponent; }
```

Comment:
Do we need to cache this in a member variable? It seems like the check is pretty cheap and only happens in failure cases.

Reply:
Ah yes, I think you right thanks :)

Reply:
I think since we are changing the approach it may make sense to just inline since its a simple logic and there is no need to reuse anymore.

Reply:
Yeah, that makes sense to me!